src/goto-programs/goto_functions.h

```diff
@@ -122,8 +123,12 @@ class goto_functionst
   ///
   /// The validation mode indicates whether well-formedness check failures are
   /// reported via DATA_INVARIANT violations or exceptions.
-  void validate(const namespacet &ns, const validation_modet vm) const
+  void validate(
+    const namespacet &ns,
+    const validation_modet vm,
+    const goto_model_validation_optionst &goto_model_validation_options) const
   {
+    validate_goto_model(*this, vm, goto_model_validation_options);
```

Comment:
Would you mind adding a comment what this is meant to do? We are inside an object that is part of any `goto_modelt`, and thus I'd expect this call to actually cause unbounded recursion?!

Reply:
Ah. No, no recursion involved. But that I should cause anyone to wonder about that is not a good thing... 

Thought to add comment 
    // validate_goto_model does a number of checks to ensure the goto_program
    // is well formed. No recursion involved as operates at the function_mapt
    // level and does not call goto_functionst::validate
    // (or any other validate methods at the goto_program level or below).

But to avoid any confusion about such recursion being possible seems best to make this call in ```
  goto_modelt::validate (
 	validate_goto_model(
goto_functions, vm, goto_model_validation_options);
goto_functions.validate(ns, vm,  goto_model_validation_options);
)
```
======================================================
Addendum.

My original line of thought:

This is in the original commit:
" NB use of validate_goto_model(*this, vm) in goto_functions.h is to avoid
circular inclusion as function_mapt is a nested type and so cannot be
forward declared."

I thought seeing this in goto_modelt::validate might be confusing/debated:
```
goto_modelt::validate (
validate_goto_model(
goto_functions, vm, goto_model_validation_options);   
goto_functions.validate(ns, vm,  goto_model_validation_options);
)
```

So I chose this next-best location
```
goto_functionst::validate(
validate_goto_model(*this, vm, goto_model_validation_options);
)

```
But I think it is far far better to avoid all doubts about recursion being possible :)